### ISSUE TITLE
Possible bug in scipy.sparse.csgraph.laplacian

### ISSUE THREAD

**Issue by snapfinger** [scipy/scipy#14490]
In [this issue](https://github.com/scipy/scipy/issues/10505) and associated [PR](https://github.com/scipy/scipy/pull/10526) it is said that the function `scipy.sparse.csgraph.laplacian` with normalization is the symmetric normalized laplacian `L_sym = D^-1/2 L D^-1/2`, where `D` is the degree matrix, `L` is the unnormalized lapalcian (`D - A`, `A` being the adjacency matrix). 

However, from the toy example below, it doesn't seem to be the case.While networkx's output matches with the `L_sym` formula's, scipy's output is quite different.

It looks like in the source code of [scipy's normalized laplacian](https://github.com/scipy/scipy/blob/44e31b1eb202ef91caad059952ba268d2610a60b/scipy/sparse/csgraph/_laplacian.py#L112-L126) there's some extra processing probably to deal with inversion of possible zeros in the diagonal, but it doesn't seem to make the result right. Following the formula, I modified the source code a bit and it can produce output mostly consistent with networkx's (my ver part below), though it definitely needs extra handling to make it robust.

#### Reproducing code example:

```python
from scipy.sparse import csgraph
import numpy as np
import networkx as nx

graph = np.array([[10, 8, 5], [8, 10, 3], [5, 3, 10]]).astype(np.float32)
print("-------------graph-----------\n", graph)

# scipy
L = csgraph.laplacian(graph, normed=True)
print("\n-------------scipy------------\n", L)

# networkx
L_nx = nx.normalized_laplacian_matrix(nx.from_numpy_array(graph)).todense()
print("\n-------------networkx:-------------\n", L_nx)

# ------------ scipy my ver
A = graph.copy()
w = A.sum(axis=0)
w = np.sqrt(w)
n, m = A.shape
diags = A.sum(axis=1).flatten()
A /= w
A /= w[:, np.newaxis]
I = np.identity(3)
my_lap = I - A
print("\n-----------my ver-------------\n", my_lap)

```

#### Output of the code above

```
-------------graph-----------
 [[10.  8.  5.]
 [ 8. 10.  3.]
 [ 5.  3. 10.]]

-------------scipy------------
 [[ 1.         -0.6689936  -0.49029034]
 [-0.6689936   1.         -0.3198011 ]
 [-0.49029034 -0.31980106  1.        ]]

-------------networkx:-------------
 [[ 0.56521739 -0.3640126  -0.24573659]
 [-0.3640126   0.52380952 -0.15430335]
 [-0.24573659 -0.15430335  0.44444444]]

-----------my ver-------------
 [[ 0.56521744 -0.36401257 -0.2457366 ]
 [-0.3640126   0.52380958 -0.15430336]
 [-0.2457366  -0.15430336  0.44444436]]

```

#### Scipy/Numpy/Python version information:

```
scipy:  1.7.0
numpy:  1.19.1
networkx:  2.5
```

**Comment by czgdp1807** [comment]
I am able to reproduce your issue on development versions of `scipy` and `netwrokx` with numpy as `1.21.1`. IMO, for similar APIs, the results from `scipy` and `networkx` should match. Though there might be reasons to use a different implementation in `scipy` as opposed to `networkx`.

cc @rgommers @fuglede

**Comment by perimosocordiae** [comment]
In this case, I believe scipy's implementation is more correct than networkx's. The key issue here is how to handle self-edges in the graph (i.e., nonzero diagonal entries).

The scipy implementation explicitly sets the diagonal to zero before computing the degrees:
https://github.com/scipy/scipy/blob/15417d73cb2a1c4bb2bb6530e30d2b566632943e/scipy/sparse/csgraph/_laplacian.py#L114

Note that in the unnormalized case (`L = D - A`) this doesn't make a difference, as the self-edge contribution to `D` is cancelled out by their presence in `A`.

**Comment by snapfinger** [comment]
I don't think (normalized) laplacian requires excluding self-edges. 

Take, for example, the def in [Luxburg's tutorial on spectral clustering](https://link.springer.com/content/pdf/10.1007/s11222-007-9033-z.pdf) (section 2.1):  `the degree of a vertex v_i \in V is defined as d_i = \sum_{j=1}^n w_{ij}` (it does not exclude `i` in the sum)

If users want to exclude self edges, we can set the diagonal of the adjacency matrix ourselves. But in many cases, it's important to keep self-edges, e.g. use graph to represent similarities between a group of people,  everyone should be most similar to themselves (correlation = 1, and this is the self edge in the adjacency matrix), with the current scipy's API, it totally changed the intended graph structure.

**Comment by lobpcg** [comment]
I agree with @snapfinger and suggest removing `np.fill_diagonal(m, 0)` adding a user warning if the original diagonal is non-zero for backward compatibility.

**Comment by lobpcg** [comment]
@snapfinger please change the title to "ENH" from "BUG". This is not a bug.

**Comment by snapfinger** [comment]
> @snapfinger please change the title to "ENH" from "BUG". This is not a bug.

Thanks, changed.

**Comment by capslockwizard** [comment]
As of Scipy 1.15.2, scipy.sparse.csgraph.laplacian does not properly implement the normalized symmetric Laplacian of the form $L_{sym} = D^{-1/2} L D^{-1/2}$. Instead, it is taking the diagonal of L and taking the square root inverse of it. See the relevant code [here](https://github.com/scipy/scipy/blob/450ef31bcbf1531d53e7ce98a9798dbde0d57d49/scipy/sparse/csgraph/_laplacian.py#L407).

**Comment by lucascolley** [comment]
cc @fernandodiazdiaz @dschult maybe you could chime in here?

**Comment by dschult** [comment]
Summary:  It looks like the normed case for csgraph.laplacian is not supported for asymmetric (directed) graphs. The docs say it is supported for the symmetric case.  Also, for the symmetric case, [the formula is](https://github.com/scipy/scipy/blob/450ef31bcbf1531d53e7ce98a9798dbde0d57d49/scipy/sparse/csgraph/_laplacian.py#L387) essentially: 
`diag[:, np.newaxis] * L * diag[: np.newaxis]`.  
But it seems to me that this should be `diag[np.newaxis, :]` for one of the two sides and `diag[:, np.newaxis]` for the other.  @lobpcg  can you check this part of the formula? How can I understand whether it is correct or not?

---

My understanding is that `graph_sum - graph_diagonal` is the standard way (at least in much of the literature... and in all the cases that I see which even consider self-loops) to exclude the self-loops from the computation of weighted degree when computing the laplacian. Also, in the symmetrized case we subtract the diagonal twice (once for row sums and once for column sums).

As for the part of the question about using matrix multiplication by $D^{-1/2}$ vs multiplying by `1/sqrt(diag)`, be careful with the type of multiplication being used. `D` is a diagonal matrix. So matrix multiplying by $D^{-1/2}$ can be described as multiplying each row (or column depending on which multiplies which) by the inverse square root of the diagonal entries of `D`.  And that is precisely the same and doing element-wise multiplication by the vector that is the reciprocal square root of the diagonal entries. The pre and post multiplication is accomplished by multiplying rows and columns IIUC.

So `np.diag(d) @ L @ np.diag(d)` is the same `d[:, np.newaxis] * L * d[np.newaxis, :]`.
For the symmetric case, it might be that the `np.newaxis` placeemnt doesn't matter -- idk -- but it does matter for the asymmetric case.

**Comment by lobpcg** [comment]
I might have been the one writing this code years ago. My code has been fully consistent with the notes in https://en.m.wikipedia.org/wiki/Laplacian_matrix
Which has not changed since then. 
To my recollection I might have even used those wikipedia examples for unit tests. 

What is exactly the question I could help with?

**Comment by lobpcg** [comment]
@dschult "How can I understand whether it is correct or not?" - it passes all unit tests 😃

**Comment by dschult** [comment]
Thanks @lobpcg for taking a look.

I'm not sure that the unit tests cover this distinction. But it currently does not seem to be a literal translation of the wikipedia page.  I have checked that the code does indeed implement those equations. I was hung up on the asymmetric (directed graph) and this code doesn't claim to handle that case.

@capslockwizard can you please clarify your concern? The released version removes self-loops from consideration when normalizing. This issue would like self-loops to be accounted for in the normalization.
